python/qemu/ot/otp/descriptor.py

```diff
@@ -118,48 +160,191 @@ def __init__(self, otpmap: 'OtpMap'):
         print(f'/* Generated from {hjname} with {scriptname} */')
         print(file=cfp)
-        print('/* clang-format off */', file=cfp)
-        for reg, off in reg_offsets:
-            print(f'REG32({reg}, {off}u)', file=cfp)
+        for slot in slots:
+            if slot.part:
+                continue
+            print(f'REG32({slot.name}, {slot.offset}u)', file=cfp)
         print(file=cfp)
-        regwidth = max(len(r[0]) for r in reg_sizes)
-        for reg, size in reg_sizes:
-            print(f'#define {reg:{regwidth}s} {size}u', file=cfp)
+
+        regwidth = max(len(s.name) for s in slots)
+        regwidth += len('_SIZE')
+        for slot in slots:
+            if slot.gen and not slot.part:
+                continue
+            name = f'{slot.name}_SIZE'
+            print(f'#define {name:{regwidth}s} {slot.size}u', file=cfp)
         print(file=cfp)
+
+        part_names = [slot.name for slot in slots if slot.part]
         pcount = len(part_names)
+        part_names = [f'OTP_PART_{pn}' for pn in part_names]
         part_names.extend((
-            '_OTP_PART_COUNT',
-            'OTP_ENTRY_DAI = _OTP_PART_COUNT',
-            'OTP_ENTRY_KDI',
-            '_OTP_ENTRY_COUNT'))
+            'OTP_PART_COUNT',
+            'OTP_ENTRY_DAI = OTP_PART_COUNT, '
+                '/* Fake partitions for error (...) */',
+            'OTP_ENTRY_KDI, /* Key derivation issue, not really OTP */',
+            'OTP_ENTRY_COUNT'))
         print('typedef enum {', file=cfp)
         for pname in part_names:
-            print(f'    {pname},', file=cfp)
+            print(f'    {pname}{"," if "," not in pname else ""}', file=cfp)
         print('} OtOTPPartitionType;', file=cfp)
         print(file=cfp)
+
         print('static const char *PART_NAMES[] = {', file=cfp)
+        print('    /* clang-format off */', file=cfp)
         for pname in part_names[:pcount]:
             print(f'    OTP_NAME_ENTRY({pname}),', file=cfp)
+        print('    /* fake partitions */', file=cfp)
+        print('    OTP_NAME_ENTRY(OTP_ENTRY_DAI),', file=cfp)
+        print('    OTP_NAME_ENTRY(OTP_ENTRY_KDI),', file=cfp)
+        print('    /* clang-format on */', file=cfp)
         print('};', file=cfp)
-        print('/* clang-format on */', file=cfp)
+        print(file=cfp)
+
+        cases: list[str] = []
+        for slot in slots:
+            if slot.part:
+                continue
+            if slot.gen:
+                cases.append(f'CASE_WIDE({slot.name});')
+            elif slot.size > 4:
+                cases.append(f'CASE_RANGE({slot.name});')
+            elif slot.size == 1:
+                cases.append(f'CASE_BYTE({slot.name});')
+            elif slot.size < 4:
+                cases.append(f'CASE_SUB({slot.name}, {slot.size}u);')
+            else:
+                cases.append(f'CASE_REG({slot.name});')
+
+        code = '''
+        static const char *ot_otp_dj_swcfg_reg_name(unsigned swreg)
```

Comment:
Could this take a `topname` arg so it can also be used for Earlgrey and any future tops?
```suggestion
        static const char *ot_otp_{top}_swcfg_reg_name(unsigned swreg)
```

Reply:
It turned out that we may not have this info, so the feature can only be made optional.

Reply:
Default to top-less naming.